# LD-annot
A bioinformatics tool to automatically provide candidate SNPs with annotations for genes in linkage disequilibrium.
# LD-annot/0.1
# Email: jprunier.1@gmail.com

LD-annot estimates experiment-specific linkage disequilibrium to delineate regions genetically linked to each genetic markers from a list of polymorphisms (most often candidate SNPs from GWAS) and provide coordinates and annotations for genes included or overlapping such regions.

## Citing LD-annot
Prunier J, Lema√ßon A, Bastien A, Mohsen, Porth I., Robert C and Droit A. submitted. LD-Annot: a bioinformatics tool to automatically provide candidate SNPs with annotations for genes in linkage disequilibrium.


## Implementation and requirements
LD-annot is meant to be easy to install and use for biologists without advanced bioinformatics skills. It simply requires to be downloaded and palced into a folder. LD-annot is actually divided into to scripts:
1) a python script (LD-annot.py) written in Python3 (not 2.7 that will not run properly and delivered a error message)
2) a bash script (calculLD.sh) that need to be in the same folder and will be invoked by LD-annot.py

The tool also needs the installation of PLINK1.9 that you may find here: https://www.cog-genomics.org/plink2

## Install

## How to run


